### PR TITLE
[Model] [Draft PR] Add support for SmallThinker model series

### PR DESCRIPTION
**Pull request by SorryMaker2022**
## Motivation

SmallThinker is a family of on-device native Mixture-of-Experts (MoE) language models specially designed for local deployment, co-developed by the IPADS and School of AI at Shanghai Jiao Tong University and Zenergize AI. Designed from the ground up for resource-constrained environments, SmallThinker brings powerful, private, and low-latency AI directly to your personal devices, without relying on the cloud.
See also:
[Huggingface](https://huggingface.co/PowerInfer/SmallThinker-21BA3B-Instruct)

This PR is to add support for the SmallThinker series of models to SGLang.

## Modifications

SmallThinker models uses ReGLU as MoE activation funcion. These modifications introduces ReGLU to FusedMoE class.
[python/sglang/srt/layers/moe/fused_moe_native.py](https://github.com/SorryMaker2022/sglang/commit/7afd7ff9dfb568049f36d6dcff30b151dbaaf8da#diff-d6daaa2727ab7438b305cb90b994e4c6ea1805a88f35eccd02e20d87cb57d515)
[python/sglang/srt/layers/moe/fused_moe_triton/fused_moe.py](https://github.com/SorryMaker2022/sglang/commit/7afd7ff9dfb568049f36d6dcff30b151dbaaf8da#diff-5382f77b5e25c8661b3473c311c2366004ad1c66e643dbd8001c1f244c74522d)
[‎python/sglang/srt/layers/moe/fused_moe_triton/triton_kernels_moe.py](https://github.com/SorryMaker2022/sglang/commit/7afd7ff9dfb568049f36d6dcff30b151dbaaf8da#diff-5205a5b8b88bbd46192c51385c5d6a990878a4e6031a4699b41d0881220e77e2)

Modeling file is introduced.
[‎python/sglang/srt/models/smallthinker.py](https://github.com/SorryMaker2022/sglang/commit/7afd7ff9dfb568049f36d6dcff30b151dbaaf8da#diff-768e07aa16214a1d9a145e5275ef51af50aab01bf0d3c7dca060cb920546a3cc)

## TODOs

The support is preliminary, users can serve SmallThinker models on the branch, but more has to be done to merge the changes to the latest SGLang.

- The branch is based on release v0.4.9 Post 2, for the FusedMoE interface is being refactored on the main branch. The code needs a rebase to v0.4.9 Post 4 after the FusedMoE interface gets fully refactored.
- SmallThinker models will be available on huggingface shortly (after a few revisions), the docs will be updated after the models get available publicly.
- The ReGLU calculation currently uses torch native implementation, I can try to introduce the fused fatrelu_and_mul fused kernels later to enhance performance.
- The model implementation only supports FusedMoE for now, I plan to support EPMoE, DeepEPMoE and DeepGEMM later.

# Test results

The test results align with the benchmarks in the model technical report or transformers test result:

| Framework | MMLU | Needle-in-a-haystack-single 16K | IFEval | MATH-500 | GPQA-Diamond |
| :---: | :---: | :---: | :---: | :---: | :---: |
| SGLang | 84.40 (-0.03) | 100 ( -0.00) | 85.58 (-0.19) | 81.40 (-1.00) | 54.55 (-0.50) |
| Transformers | - | 100 | - | - | - |
| Technical Report | 84.43 | - | 85.77 | 82.40 | 55.05 |

Test settings: `python -m sglang.launch_server --model-path PowerInfer/SmallThinker-21BA3B-Instruct --trust-remote-code --enable-torch-compile --dp 4 --tp 2 --cuda-graph-max-bs 32`

Evaluation platform: opencompass

The inference speed of 21BA3B model on a H200 GPU is 208.06 tokens/s, which aligns with Qwen3 30B-A3B at a similar activation scale. The inference speed of 4BA0.6B model is 441.24 tokens/s and 218.9 tokens/s on a H200 GPU and a 4090 GPU respectively.

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/models/smallthinker.py [hunk 1]
```diff
@@ -0,0 +1,595 @@
...
+        self.num_active_primary_experts = config.moe_num_active_primary_experts
+        self.moe_primary_router_apply_softmax = config.moe_primary_router_apply_softmax
+        self.tp_size = get_tensor_model_parallel_world_size()
+
+        # Primary router
+        self.primary_router = ReplicatedLinear(self.hidden_dim, self.num_primary_experts, bias=False, skip_bias_add=False, 
+                                               quant_config=quant_config, prefix=f"{prefix}.primary_router")
+
+        def custom_topk(hidden_states, gating_output, topk, renormalize):
+            router_logits, selected_experts = torch.topk(gating_output[0], topk, dim=-1)
```
[the remark this answers is not in the document]
Comment (SorryMaker2022) [reply python/sglang/srt/models/smallthinker.py:67]: Replicator returns a tuple, with its first element the gating_output tensor we want. This is NOT a bug.

File: python/sglang/srt/models/smallthinker.py [hunk 2]
```diff
@@ -0,0 +1,595 @@
...
+
+        hidden_states = self.self_attn(
+            positions=positions,
+            hidden_states=hidden_states,
+            forward_batch=forward_batch,
+        )
+        
+        # MLP
+        hidden_states, residual = self.post_attention_layernorm(hidden_states, residual)
+        hidden_states, _ = self.mlp(router_input, hidden_states, forward_batch)
```
[the remark this answers is not in the document]
Comment (SorryMaker2022) [reply python/sglang/srt/models/smallthinker.py:274]: This is the technical feature of SmallThinker models. SmallThinker models use the hidden_states before attention to calculate routing weights, but use the output of attention to calculate the results of the MLP. It is intended to be there.
